File: packages/node-utils/src/lib/exec.ts
import { exec as _exec } from 'child_process';
import { promisify } from 'util';
import { log } from "utils";

const execAsync = promisify(_exec);

//
// Executes a command using the specified tool.
//
export function exec(command: string): Promise<{ stdout: string; stderr: string }> {
    return execAsync(command);
}

//
// Runs the command and adds logging for the tool used.
//
export async function execLogged(tool: string, command: string, validate?: () => Promise<string | undefined>): Promise<{ stdout: string; stderr: string }> {
    log.verbose(`Executing ${tool} with command: "${command}"`);
    try {
        const { stdout, stderr } = await execAsync(command);
        log.tool(tool, { stdout, stderr });
        if (validate) {
            const validationFailedReason = await validate();
            if (validationFailedReason) {
                const msg = `Validation failed for command: ${command}\nReason: ${validationFailedReason}`;
                log.error(msg);
                log.info(`===\nCommand: ${command}\nCommand stdout:\n${stdout ? stdout : 'No output'}\nCommand stderr:\n${stderr ? stderr : 'No error'}\n===`);
                throw new Error(msg);
            }
        }
        return { stdout, stderr };
    }
    catch (error: any) {
        const msg = `Failed to execute command: ${command}`;
        log.exception(msg, error);
        throw new Error(msg);
    }
}

File: packages-zig/node-utils-zig/src/lib/exec.zig
const std = @import("std");
const builtin = @import("builtin");
const utils = @import("utils-zig");
const errors = utils.errors;
const process_env = @import("process-env.zig");

//
// The output of a command run with exec (TypeScript: `{ stdout: string; stderr: string }`).
//
pub const ExecResult = struct {
    // Everything the command wrote to stdout.
    stdout: []const u8,

    // Everything the command wrote to stderr.
    stderr: []const u8,
};

//
// Executes a command using the specified tool.
// Like Node's `child_process.exec`, the command runs in the shell (/bin/sh -c, or cmd.exe on Windows)
// with the environment of `process.env` (see process-env.zig), and fails when the command cannot be
// started or exits with a non-zero code. The error message is Node's: "Command failed: <command>\n<stderr>".
//
pub fn exec(allocator: std.mem.Allocator, io: std.Io, command: []const u8) !ExecResult {
    const argv: []const []const u8 = if (builtin.os.tag == .windows)
        &.{ "cmd.exe", "/d", "/s", "/c", command }
    else
        &.{ "/bin/sh", "-c", command };
    const result = std.process.run(allocator, io, .{
        .argv = argv,
        .environ_map = process_env.getEnvironMap(),
    }) catch |err| {
        return errors.throwError("Command failed: {s}\n{s}", .{ command, @errorName(err) });
    };
    const succeeded = switch (result.term) {
        .exited => |code| code == 0,
        else => false,
    };
    if (!succeeded) {
        return errors.throwError("Command failed: {s}\n{s}", .{ command, result.stderr });
    }
    return .{ .stdout = result.stdout, .stderr = result.stderr };
}

// Not ported: execLogged (not used by replicate or verify).
